experiments/targeting.go

```diff
@@ -0,0 +1,345 @@
+package experiments
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// Targeting is the common interface to implement experiment targeting.
+// Evaluated whether the provided input matches the expected values.
+type Targeting interface {
+	Evaluate(inputs map[string]interface{}) bool
+}
+
+// NewTargeting parses the given targeting configuration into a Targeting.
+func NewTargeting(targetingConfig []byte) (Targeting, error) {
+	var config map[string]interface{}
+	decoder := json.NewDecoder(bytes.NewBuffer(targetingConfig))
+	// ensures numbers are parsed into json.Number in order to differentiate
+	// between integer and float64 later on
+	decoder.UseNumber()
+	err := decoder.Decode(&config)
+	if err != nil {
+		return nil, err
+	}
+	if len(config) != 1 {
+		return nil, TargetingNodeError("call to create targeting tree expects single input key")
+	}
+	return parseNodes(config)
+}
+
+func parseNodes(config map[string]interface{}) (Targeting, error) {
+	key := firstKeyStr(config)
+	return mapOperatorNode(strings.ToLower(key), config[key])
+}
+
+func parseNode(node interface{}) (Targeting, error) {
+	switch n := node.(type) {
+	case []interface{}:
+	case map[interface{}]interface{}:
+		key := firstKey(n)
+		return mapOperatorNode(strings.ToLower(key), n[key])
+	case map[string]interface{}:
+		key := firstKeyStr(n)
+		return mapOperatorNode(strings.ToLower(key), n[key])
+	}
+	return nil, TargetingNodeError(fmt.Sprintf("node type %T unknown", node))
+}
+
+func firstKeyStr(m map[string]interface{}) string {
```

Comment:
Is it that when we use `firstKeyStr` and `firstKey`, the map always contain exactly one key-value pair? If that's not the case, the iteration order of maps are intentionally randomized in go runtime so that you will usually get different `firstKeyStr`/`firstKey` results for the same map.

Reply:
Yes, the maps will always only have one key of the way experiments configurations are structured. Here is an example

Reply:
It's possible to define a struct with `Key` and `Value` (or whatever names), then in it's `UnmarshalJSON` implementation, try to unmarshal to a map, then check that map has len of 1, then get the key and value out.

Reply:
Something like this: https://play.golang.com/p/LCGwucidXXV

Reply:
Oh that's good. Do you think it's worthwhile changing it to use? I think the only problem is that it would still need to be decided at the stage of looking at the key which concrete struct to set it up with, i.e. `AnyNode`, `AllNode`, etc.

Reply:
I think it makes more sense to use that if we get rid of the `map[interface{}]interface{}` case, but since this PR is already getting big we can do that in a followup PR instead.